compiler/rustc_middle/src/mir/mod.rs

```diff
@@ -2267,6 +2270,16 @@ pub enum Rvalue<'tcx> {
     /// away after type-checking and before lowering.
     Aggregate(Box<AggregateKind<'tcx>>, Vec<Operand<'tcx>>),
 
+    /// Creates a value of the variant with all fields uninitialized
+    ///
+    /// Specifically, this has the property that once all fields of the variant are also
+    /// initialized, the memory of the place this is assigned to is a valid value for the type, and
+    /// that the `discriminant` rvalue for that value will yield the discriminant of the variant
+    /// here.
+    ///
+    /// Only allowed for ADTs.
+    UninitVariant(Ty<'tcx>, VariantIdx),
```

Comment:
Why not have a "full wipe" `dest = Uninit` before starting to initialize the fields?

Also, not sure what purpose automatic initialization serves - is the goal to remove `SetDiscriminant`?

`SetDiscriminant` is still needed to "bless" the ADT value (think partial initialization vs full value initialization wrt borrowck and whether the destructor is active). I don't feel like #94590 provides sufficient motivation - it seems to *assume* `SetDiscriminant` is wrong (it's not) and ignore its current (intended) semantics.

Perhaps the name was poorly chosen, maybe it should've been `FinalizeVariant`?

Reply:
cc @nikomatsakis

Reply:
> Why not have a "full wipe" `dest = Uninit` before starting to initialize the fields?

Yeah, I should make the documentation more clear about this. How about saying something like. This rvalue is created by starting with all uninit bytes and then possibly writing to some of the bytes if the ADT variant requires some bytes to be set. All fields are guaranteed to remain uninit.

In other words, I feel like we are already doing something similar to this.

> Also, not sure what purpose automatic initialization serves - is the goal to remove `SetDiscriminant`?
> 
> `SetDiscriminant` is still needed to "bless" the ADT value (think partial initialization vs full value initialization wrt borrowck and whether the destructor is active). I don't feel like #94590 provides sufficient motivation - it seems to _assume_ `SetDiscriminant` is wrong (it's not) and ignore its current (intended) semantics.
> 
> Perhaps the name was poorly chosen, maybe it should've been `FinalizeVariant`?

So, the borrowck and thing about destructors is a complete non-issue imo because `SetDiscriminant` does not exist in that part of MIR. It is added during deaggregation and generator lowering, which runs after borrowck and drop elaboration.

This is getting a bit out of scope, but I'd like to argue that generally, MIR falls into two categories:
 1. "Early" MIR - this is the MIR roughly up to drop elaboration/deaggregation. In this part of MIR, we are usually attempting to prove the correctness of some parts of the program - borrowck and drop elaboration both do this kind of analysis. In such cases, it is important to be able to answer the question "is a place fully initialized," because if the answer to this is "no" then we might have to return an error.
 2.  "Late" MIR - this is MIR that begins after deaggregation (or somewhere around there). In late MIR, we are no longer attempting to do static analysis to prove some things about the correctness of the program - instead, we are doing optimizations. This is both typical optimizations, but also generator lowering (much of the complexity of which seems to stem from it trying to be smart). In late MIR, we - as far as I know - never have to prove that a place is initialized. We actually get to assume it, because early MIR has already proved the program "correct" and contains no UB. Instead, optimizations are generally interested in proving other things, like that a place is dead.

The `UninitVariant` rvalue is very clearly tailored towards the second category (and that's good, because that's the only place it appears). It makes proving that a place is dead very easy - it makes proving that a place is harder.

Reply:
Oh, I realize now that you meant in addition to the original `SetDiscirminant`, not instead of. That's definitely possible, but the question of the operational semantics of `SetDiscriminant` remains. If it asserts validity of the whole value (which is what I am understanding you want), I still think that causes issues. It makes it illegal to DSE this for example:
```
_1 = uninit;
(_1 as V).0 = val;
(_1 as V).1 = val2;
SetDiscriminant(_1) = V;
(_1 as V).0 = val3;
```
And at the same time, I don't really see what benefits it brings

Reply:
Maybe the issue is that SetDiscriminant is overloaded with setting the discriminant and marking the value as initialized. We could split this operation into two statements. Then we could have something to deinit the place, something to set the discriminant and something to mark the place as fully initialized (the latter invoking miri's validity checks)

Strawman proposal:

1. Reset value by using StorageDead followed immediately by StorageLive
2. SetDiscriminant right where it is right now
3. Assign from the place to itself, and make such assignments special in the sense that they don't do anything beyond guarantee validity

Reply:
Hmm... yea, that's exactly what eddyb proposed...

Reply:
I'm still not understanding the use for such an assertion. The assertion can never fail in MIRI for the output of deaggregation (and unoptimized MIR in general, as far as I know). Meanwhile, for optimizations it seems more of a hazard than an asset - maybe that just shows I don't know enough, but I still would like to see a motivating example.

I would also be happy with the `StorageDead` followed by `StorageLive`, but I don't see how it's possible given that the LHS might not be a local

Reply:
In order for deaggregation to be non-lossy, don't we need a finalize statement? Otherwise we have to guess whether reaggregating is ok depending on whether the type has diverging fields

But I think we can just change the rules for SetDiscriminant right after we have the uninit rvalue. SetDiscriminant would then become literally just what it describes, and when we move deaggregation earlier we add the finalize thing. 

We could even stop emitting SetDiscriminant for enums where it would bea no-op

Reply:
> In order for deaggregation to be non-lossy, don't we need a finalize statement? Otherwise we have to guess whether reaggregating is ok depending on whether the type has diverging fields

What are diverging fields?

Reply:
Consider the following struct:

```rust
struct Foo<T>(u32, T);
```

We can soundly know that a value of `Foo<bool>` is initialized, by looking at the field assignments. So just from that perspective, the following code could be legal:

```rust
let x: Foo<bool>;
x.0 = 42;
x.1 = true;
```

Now, when you have a generic functions, you could be doing something funky like

```rust

Reply:
@oli-obk did you post the comment too early?

Reply:
I don't like conflating the action of setting a discriminant with the "analysis marker" of "all fields have been initialized". Those seem like entirely orthogonal concepts to me. If we need such a marker because otherwise the analysis is too hard to get right, IMO it should be basically a NOP (in Miri it might just check the validity invariant in-place and that's it).

It also seems odd (as before) that generators and enums are treated so differently. But that could be resolved by renaming `UninitVariant` to just `Uninit`, which is strictly more general -- one can simply call `SetDiscriminant` after `x = Uninit`. (I am not sure when `Uninit` would even be generated; my concern here is solely MIR and its semantics.)

@nikomatsakis  The issue is less defining a formal semantics for these operations and more figuring out which are the consumers of the IR and which primitives best fit their needs. And indeed we have some conflict of "analysis" (checking whether variables are initialized) and "optimizations" here.

That said, we do have `Rvalue::Aggregate`, which seems just right to me in terms of an "analysis MIR" -- if only it also supported enums. Deaggregation would then correspond to translating from analysis MIR to optimization MIR, and for the latter we don't need any clearly defined notion of "this is fully initialized" any more.

So, that makes me wonder -- @oli-obk why do you want to move deaggregation into "analysis MIR"? That seems counter-productive to me.

Reply:
@oli-obk I'm not sure I follow that example: When `Foo<T>` is deaggregated for any `T` (including a generic `T`) it produces 
```rust
P.0 = val1;
P.1 = val2;
```
That is sound to re-aggregate for all `T` - if `T` is `!` (or some other uninhabited type) then both the deaggregated and re-aggregated version are UB.

Now, if some pass wanted to be smarter about this and realize that the `P.1` assignment is unnecessary in eg `Foo<()>` and so
```rust
P.0 = val1;
```
can be re-aggregated, it could do so. However, that pass would introduce a new operand of type `T` that wasn't there before, and it needs to ensure that that change is sound - this is where the `Foo<!>` re-aggregation would fail.

Reply:
@RalfJung in general, :+1: to everything you said. That being said, I would like to argue that generators and enums actually are already difficult in an important way: Rust allows us to change the discriminant (or "active variant" to the extent such a thing exists) of a generator in a way that allows other data to be preserved. Specifically, we can do something like this:
```rust
let mut x = 1;
foo.await;
dbg!(x);
x = 2;
bar.await;
dbg!(x);
```
`x` is present as a field in the variants corresponding to both `.await`s - and this does indeed have to be the same field because there might be references to it and we need address stability. Enums have no analogue to this, which is why `SetDiscriminant` always felt a little bit unnatural to me. It does not correspond to an operation that you can do in Rust outside of the result of deaggregation

Reply:
Yes, I see that. But I feel like `SetDiscriminant` is a reasonable primitive useful in both situations.

Basically, the hallmark of an expressive and compositional language is that you can put together the building blocks in different ways to get different results. In contrast, having a separate primitive for each new thing is the opposite of compositional. That's why `SetDiscriminant` feels to me like better language design than separate handling of enum and generator discriminants.

Reply:
Maybe, but I also think it's equally important that you can't use the building blocks to express things you're not supposed to be able to express. `SetDiscriminant` allows for code like this (with layout checks):
```rust
enum E {
    A(u8),
    B(u8),
}

(_1 as A).0 = 1;
SetDiscriminant(_1) = A;
// use `_1`
SetDiscriminant(_1) = B;
_2 = (_1 as B).0;
```

Both the alternatives I've proposed cleanly prevent writing such MIR.

That being said, I don't actually have particularly strong opinions on this anymore. The things I am most concerned about are the de-coupling of `SetDiscriminant`s analysis and operational roles, and ensuring that deaggregation is not lossy

Reply:
> Maybe, but I also think it's equally important that you can't use the building blocks to express things you're not supposed to be able to express.

I guess the question is if you are truly not supposed to be able to express this. We need it for generators after all, so why not for enums?

Reply:
@nikomatsakis that sounds good, I'd be happy to do such a writeup. I'm not familiar with the format you're referring to - is there somethijg more specific that you're imagining than a hackmd detailing the concerns and options?